src/QRCoders.jl

```diff
@@ -101,6 +101,79 @@ Error correction level that can restore up to 30% of missing codewords.
 """
 struct High <: ErrCorrLevel end
 
+"""
+    QRCode
+
+A type that represents a QR code.
+
+# Fields
+- `version::Int`: version of the QR code
+- `mode::Mode`: encoding mode of the QR code
+- `eclevel::ErrCorrLevel`: error correction level of the QR code
+- `mask::Int`: mask pattern of the QR code
+- `message::AbstractString`: message to be encoded
+- `width::Int`: width of the white border
+"""
+mutable struct QRCode
+    version::Int # version of the QR code
+    mode::Mode # encoding mode
+    eclevel::ErrCorrLevel
+    mask::Int # mask pattern
+    message::AbstractString # message to be encoded
+    width::Int # width of the white border
+end
```

Comment:
Making it a mutable struct is okay. But we'll need to ensure that this is only used at the interface level -- before the core computation, you'll need to unpack or create a more "static" version for Julia compiler to optimize it.

By "unpack", I mean something like this:

```julia
function qrcode(c::QRCode)
    version, mode, eclevel, mask, message, width = c
    # then use it
end
```

Also, I'm not sure how much overhead it is with `mode::Mode`, `eclevel::ErrCorrLevel`, `message::AbstractString` being abstract field. If it's a concern, then we might need to parameterize it into something like:

```julia
mutable struct QRCode{M<:Mode, ...}
    ...
    mode::M
    ...
end
```

Reply:
About the abstract type, I tested that in `Poly` before, and get a bad performance.
```julia
struct Poly1{T} mat::Vector{T} end
struct Poly2 mat::Vector{Int} end
struct Poly3 mat::AbstractVector{Int} end
mat = rand(1:255, 1000)
mat2 = @view(mat[1:end-1])
@btime Poly1{Int}($mat2);
@btime Poly2($mat2);
@btime Poly3($mat2);
# results
# 319.907 ns (1 allocation: 7.94 KiB)
# 304.673 ns (1 allocation: 7.94 KiB)
# 4.700 ns (1 allocation: 48 bytes)
```
`Poly3` use abstract type to avoid type convert, so it looks faster in this case. However, the overall performance is decreased due to the overhead you mentioned.

In our case, keywords in `QRCode` are not involved in calculation most of the time, they are used in cases like `f(xxx, ::Numeric)`, so this might have little effect on the performance.
More importantly, using `{M<:Mode, E<:ErrCorrLevel}` could in fact make `mode, eclevel` "immutable". I get an error when typing `code.mode = Numeric()`.
```julia
MethodError: Cannot `convert` an object of type Byte to an object of type Numeric
```

As a result, the keywords `mode, eclevel` remain unchanged in the new commit.